### PR TITLE
Drive blueprints off of a DataStore

### PR DESCRIPTION
**Pull request by jleibs**
### What
 - Uses the new serde-helper from: https://github.com/rerun-io/rerun/pull/2004 to stick space-views in the data-store
 - Refactors the frame loop to follow this basic logic:
   - Materialize the blueprint from the store
   - Save a snapshot of the materialized blueprint
   - Run most of the legacy code as is
   - If the blueprint has been modified then save the modifications back to the store
 - In the internals of the Python, this introduces a new `global_blueprint_stream()`
 - Adds a few python APIs that send data to the stream.
 - RecordingId is now a string, and we use a special condition of RecordingId == AppId to determine that a blueprint is the "default blueprint" for an app.
 - The default behavior of rr.init() now uses this special recording-id, which means in the common case your blueprint API calls do almost exactly what you want, but an expert can still opt out using `add_to_app_default_blueprint=False` (this might need a better name), in which case they get complete control over a new blueprint.
 - SpaceViewIds generated by the app use a hash of the spaceview name to avoid repeated appends.
 - The "selected blueprint" is determined based on app-id. There is a "currently selected" blueprint for each app, which defaults to the special global blueprint.

The new APIs are now gated within the `rr.experimental` module and the flag on init `exp_init_blueprint` now defauts to false.
 
 There are lots of edge cases that still need to get worked out and a fair number of TODOs:
   - There are a few pieces of dynamic app state that are somewhat inappropriately stored in the blueprint so they now get cleared on every frame. Most of these have been worked around but it all feels quite fragile until we do some more state-refactoring.
   - Any change (e.g. view orientation) gratuitoulsy re-saves the whole spaceview, so we churn through memory faster than we should.
   - Any kind of blueprint state persistence on app-termination is gone
   - Any existing blueprint you had gets blown away when you connect a new session that inits a blueprint stream
   - Switching of selected recording/blueprint and how that relates to app-id makes it very easy to get confused as to where your data/blueprint went.

Followup work:
 - https://github.com/rerun-io/rerun/issues/2158

### Experimentations:

For the notebook to work you'll need to build with web-viewer support:
```
just py-build --features web_viewer
```

There's a sample notebook in `examples/python/notebook/blueprint.ipynb`

image

Additionally for the non-notebook use-case, try running each of the following multiple times.
```
python examples/python/blueprint/main.py

python examples/python/blueprint/main.py --append

python examples/python/blueprint/main.py --skip-blueprint

python examples/python/blueprint/main.py --append --skip-blueprint
```

TODO:
 - [x] Zoom on images appears to be broken (probably depends on non-serialized state).
 - [x] Move APIs into `experiments`
 - [x] Simplify the set_panel API
 - [x] Add missing python docstrings

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)

PR Build Summary: https://build.rerun.io/pr/2010

### CODE REVIEW

File: crates/re_log_types/src/serde_field.rs
Comment (jleibs) [inline crates/re_log_types/src/serde_field.rs:1]: This is cherry-picked in from: https://github.com/rerun-io/rerun/pull/2004
Will go away after rebase/merging.

**Comment by Wumpf** [comment]
wget doesn't work from the notebook on Mac (notebook in vscode), kinda suprising. (Working around it for testing)

File: crates/re_viewer_context/src/lib.rs [hunk 1]
```diff
@@ -52,11 +52,33 @@ impl SpaceViewId {
...
+        Self(uuid)
+    }
+
     pub fn gpu_readback_id(self) -> re_renderer::GpuReadbackIdentifier {
         re_log_types::hash::Hash64::hash(self).hash64()
     }
 }
 
+impl std::fmt::Display for SpaceViewId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Comment (emilk) [inline crates/re_viewer_context/src/lib.rs:77]:
```suggestion
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

File: crates/re_viewer_context/src/lib.rs [hunk 2]
```diff
@@ -52,11 +52,33 @@ impl SpaceViewId {
...
     }
 
+    pub fn random_from_str(s: &str) -> Self {
+        // TODO(jleibs): This is overkill; we can probably get away with 64-bit ids.
+        use rand::{Rng as _, SeedableRng as _};
+        use std::hash::{Hash as _, Hasher as _};
+
+        let salt: u64 = 0x307b_e149_0a3a_5552;
+
+        let mut hasher = std::collections::hash_map::DefaultHasher::default();
```
Comment (emilk) [inline crates/re_viewer_context/src/lib.rs:62]: It's worth noting that this hash can change when we update Rust versions.

We use `ahash::RandomState` in many other places, though that can also change when we update `ahash`. However, new rust versions is outside our control, while ahash updates is something we always opt-in to.

TLDR: using `ahash::RandomState::with_seeds(1, 2, 3, 4)` instead here should be a bit better

File: crates/re_data_store/src/log_db.rs [hunk 3]
```diff
@@ -69,7 +69,8 @@ impl EntityDb {
         Ok(())
     }
 
-    fn try_add_data_row(&mut self, row: &DataRow) -> Result<(), Error> {
+    // TODO(jleibs): If this shouldn't be public, chain together other setters
+    pub fn try_add_data_row(&mut self, row: &DataRow) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_data_store/src/log_db.rs:73]: Wanted to chat with @teh-cmc about whether this should be public... if it's not there's probably some other way of adding a data-row but it requires using other less direct APIs.

File: crates/re_log_types/src/lib.rs [hunk 4]
```diff
@@ -222,6 +230,14 @@ pub struct BeginRecordingMsg {
     pub info: RecordingInfo,
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub enum RecordingType {
+    Unknown,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_log_types/src/lib.rs:258]: Yes, since it's theoretically possible to have a LogDb entry or Recording that didn't send a RecordingInfo.

File: crates/re_sdk/src/recording_stream.rs [hunk 5]
```diff
@@ -1077,6 +1087,21 @@ mod tests {
...
+            .enabled(true)
+            .batcher_config(DataTableBatcherConfig::ALWAYS)
+            .buffered()
+            .unwrap();
+
+        let storage = rec_stream.memory();
+        let msgs = storage.take();
+
+        // Nothing should have been sent if we didn't log anything
+        assert!(msgs.is_empty());
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_sdk/src/recording_stream.rs:1102]: Yes; I thought I stripped out these changes. Maybe the test was in a separate commit. Good catch.

File: crates/re_viewer_context/src/lib.rs [hunk 6]
```diff
@@ -52,11 +52,33 @@ impl SpaceViewId {
...
+        salt.hash(&mut hasher);
+        s.hash(&mut hasher);
+        let mut rng = rand::rngs::StdRng::seed_from_u64(hasher.finish());
+        let uuid = uuid::Builder::from_random_bytes(rng.gen()).into_uuid();
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer_context/src/lib.rs:66]: Good catch; I'll be sure to let this guy know since I copied the code from him:
image

File: crates/re_viewer/src/ui/blueprint_sync.rs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    space_view::SpaceViewComponent,
+    viewport::{
+        AutoSpaceViews, SpaceViewMaximized, SpaceViewVisibility, ViewportLayout, VIEWPORT_PATH,
+    },
+};
+
+use super::{Blueprint, SpaceView, Viewport};
+
+// TODO(jleibs): Put this somewhere more global
+pub fn store_one_component<C: SerializableComponent>(
```
Comment (emilk) [inline crates/re_viewer/src/ui/blueprint_sync.rs:19]: Seems like `re_query` is a good place, despite its name, since it is a mirror image of `query_single_component`

File: crates/re_viewer/src/ui/blueprint_sync.rs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+            }
+        }
+
+        // Remove any deleted space views
+        for space_view_id in snapshot.viewport.space_view_ids() {
+            if self.viewport.space_view(space_view_id).is_none() {
+                clear_space_view(blueprint_db, space_view_id);
+            }
+        }
+    }
```
Comment (emilk) [inline crates/re_viewer/src/ui/blueprint_sync.rs:84]: doing all this manually won't scale. I wonder if we should use something like https://crates.io/crates/diffus

File: crates/re_viewer/src/ui/data_blueprint.rs [hunk 9]
```diff
@@ -71,6 +71,20 @@ pub struct DataBlueprintTree {
     data_blueprints: DataBlueprints,
 }
 
+// Manually implement PartialEq since slotmap doesn't
+impl PartialEq for DataBlueprintTree {
+    fn eq(&self, other: &Self) -> bool {
+        self.groups
```
Comment (emilk) [inline crates/re_viewer/src/ui/data_blueprint.rs:77]: You should use the exhaustive `let Self { groups, path_to_group, … } = self` destruction trick here to make sure this gets updated when `DataBlueprintTree` changes

File: crates/re_viewer/src/ui/view_spatial/ui.rs [hunk 10]
```diff
@@ -92,6 +92,23 @@ pub struct ViewSpatialState {
     auto_size_config: re_renderer::AutoSizeConfig,
 }
 
+// TODO(jleibs): This render-related state probably doesn't belong in the blueprint
+// in the blueprint in the first place. But since serde skips it we also have to ignore it.
+// or else we re-store state on every frame. Either way the fact that we don't get it
+// back out of the store is going to cause problems.
```
Comment (emilk) [inline crates/re_viewer/src/ui/view_spatial/ui.rs:98]: yeah, this is quite annoying. we need to figure something out

File: crates/re_viewer/src/ui/view_tensor/ui.rs [hunk 11]
```diff
@@ -44,6 +44,15 @@ pub struct ViewTensorState {
     tensor: Option<DecodedTensor>,
 }
 
+// Manually implement PartialEq so we aren't comparing the tensor
```
Comment (emilk) [inline crates/re_viewer/src/ui/view_tensor/ui.rs:47]: This is the third time now we manually skip things in `PartialEq`. Perhaps we should make them all link to the same github issue so we can find all instances quickly if/when we find a fix for it

File: crates/re_viewer/src/ui/viewport.rs [hunk 12]
```diff
@@ -22,47 +22,49 @@ use super::{
...
     /// Set to `true` the first time the user messes around with the viewport blueprint.
-    /// Before this is set we automatically add new spaces to the viewport
-    /// when they show up in the data.
-    has_been_user_edited: bool,
+    pub(crate) has_been_user_edited: bool,
 
-    #[serde(skip)]
-    space_view_entity_window: Option<SpaceViewEntityPicker>,
+    /// Whether or not space views should be created automatically.
+    pub(crate) auto_space_views: bool,
```
Comment (emilk) [inline crates/re_viewer/src/ui/viewport.rs:51]: making everything `pub(crate)` feels a bit icky

File: crates/re_viewer/src/app.rs [hunk 13]
```diff
@@ -499,32 +494,70 @@ impl eframe::App for App {
...
+                            recording_source: re_log_types::RecordingSource::Other(
+                                "viewer".to_owned(),
+                            ),
+                            recording_type: RecordingType::Blueprint,
+                        },
+                    });
+
+                    blueprint_db
+                });
+            Blueprint::from_db(egui_ctx, blueprint_db)
```
Comment (emilk) [inline crates/re_viewer/src/app.rs:530]: it would be nice to extract this to a separate function. The containing function is already pretty long

File: crates/re_viewer/src/app.rs [hunk 14]
```diff
@@ -499,32 +494,70 @@ impl eframe::App for App {
...
+                .or_insert_with(|| {
+                    let mut blueprint_db = LogDb::default();
+
+                    blueprint_db.add_begin_recording_msg(&BeginRecordingMsg {
+                        row_id: re_log_types::RowId::random(),
+                        info: re_log_types::RecordingInfo {
+                            application_id: this_frame_blueprint_id.as_str().into(),
+                            recording_id: self.state.selected_blueprint_id.clone(),
+                            is_official_example: false,
+                            started: re_log_types::Time::from_seconds_since_epoch(0.0),
```
Comment (emilk) [inline crates/re_viewer/src/app.rs:520]:
```suggestion
                            started: re_log_types::Time::now(),
```

seems more accurate?

File: crates/re_viewer/src/app.rs [hunk 15]
```diff
@@ -572,12 +606,19 @@ impl eframe::App for App {
...
         self.frame_time_history.add(
             egui_ctx.input(|i| i.time),
             frame_start.elapsed().as_secs_f32(),
         );
+
+        // If this wasn't the default blueprint, save it back
+        if this_frame_blueprint_id != Default::default() {
+            let blueprint_db = self.log_dbs.get_mut(&this_frame_blueprint_id).unwrap();
+
+            blueprint.process_updates(&blueprint_snapshot, blueprint_db);
```
Comment (emilk) [inline crates/re_viewer/src/app.rs:620]:
```suggestion
            blueprint.store_changes(&blueprint_snapshot, blueprint_db);
```

or something would be a more descriptive name imho

**Review by emilk — COMMENTED** [review]
I'm gonna leave `rerun_py/src/python_bridge.rs` to @teh-cmc

File: crates/re_viewer/src/ui/blueprint_sync.rs (hunk 8 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/blueprint_sync.rs:84]: Agreed on the scaling poorly. I think there are two main approaches to consider:
 - Editing the materialized object and re-syncing, as we're doing here, in which case something like diffus seems like a good candidate.
 - Effectively making the materialized view read-only, and instead changing the places where we modify state to instead write back changes directly to the store.

File: crates/re_viewer/src/ui/blueprint_load.rs [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    let viewport_layout: ViewportLayout = query_latest_single::<ViewportLayout>(
+        &blueprint_db.entity_db,
+        &VIEWPORT_PATH.into(),
+        &query,
+    )
+    .unwrap_or_default();
+
+    // TODO(jleibs): Can this be done as a partition operation without the clone?
+    let mut known_space_views = space_views.clone();
+    known_space_views.retain(|k, _| viewport_layout.space_view_keys.contains(k));
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer/src/ui/blueprint_load.rs:109]: Almost... except cloned doesn't work on iterator from a map type.

Needed:
```
    let unknown_space_views: HashMap<_, _> = space_views
        .iter()
        .filter(|(k, _)| !viewport_layout.space_view_keys.contains(k))
        .map(|(k, v)| (*k, v.clone()))
        .collect();
```

File: crates/re_viewer/src/ui/blueprint_sync.rs (hunk 7 above)
Comment (jleibs) [reply crates/re_viewer/src/ui/blueprint_sync.rs:19]: I actually put both in re_data_store util since they don't have a dep on re_query.

File: crates/re_viewer/src/app.rs (hunk 14 above)
Comment (jleibs) [reply crates/re_viewer/src/app.rs]: This seemed like such a reasonable suggestion until several hours later I was 3 PRs deep fighting with cross-platform time::now support :laughing: 
 - https://github.com/rerun-io/rerun/pull/2090
 - https://github.com/rerun-io/rerun/pull/2093
 - https://github.com/rerun-io/rerun/pull/2094

**Comment by teh-cmc** [comment]
The following usage pattern deterministically crashes:

https://github.com/rerun-io/rerun/assets/2910679/f513b264-a6e1-4098-bd66-9e09313b74a9

```
thread 'ThreadId(1)' panicked at 'called `Option::unwrap()` on a `None` value', arrow2_convert-0.5.0/src/deserialize.rs:33

   6: core::panicking::panic_fmt
             at core/src/panicking.rs:64:14
   7: core::panicking::panic
             at core/src/panicking.rs:114:5
   8: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at core/src/option.rs:942:21
   9: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at core/src/option.rs:1091:29
  10: re_data_store::util::query_latest_single
             at re_data_store/src/util.rs:35:21
  11: re_viewer::ui::blueprint_load::load_viewport
             at re_data_store/src/util.rs:56:5
  12: re_viewer::ui::blueprint_load::<impl re_viewer::ui::blueprint::Blueprint>::from_db
             at re_viewer/src/ui/blueprint_load.rs:37:24
  13: <re_viewer::app::App as eframe::epi::App>::update
             at re_viewer/src/app.rs:448:13
  14: egui::context::Context::run
             at eframe/src/native/epi_integration.rs:494:13
  15: eframe::native::epi_integration::EpiIntegration::update
             at eframe/src/native/epi_integration.rs:492:27
  16: <eframe::native::run::wgpu_integration::WgpuWinitApp as eframe::native::run::WinitApp>::run_ui_and_paint
             at eframe/src/native/run.rs:1276:21
  17: eframe::native::run::run_and_return::{{closure}}
             at eframe/src/native/run.rs:152:17
  18: winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run_return
  19: eframe::native::run::run_and_return
             at winit-0.28.1/src/platform_impl/linux/mod.rs:785:56
  20: std::thread::local::LocalKey<T>::with
             at eframe/src/native/run.rs:1468:17
  21: eframe::native::run::wgpu_integration::run_wgpu
             at eframe/src/native/run.rs:109:5
  22: eframe::run_native
             at eframe/src/lib.rs:167:13
```

**Comment by teh-cmc** [comment]
Resetting the viewer now puts us in a state where no blueprint is selected, which is pretty confusing:

https://github.com/rerun-io/rerun/assets/2910679/fb375021-3cf8-4330-83af-77dfacac58e7

I would expect it to fall back to the default blueprint?

File: examples/python/blueprint/main.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    )
+    parser.add_argument("--auto-space-views", action="store_true", help="Automatically add space views")
+
+    args = parser.parse_args()
+
+    if args.blueprint_only:
+        # If only using blueprint, it's important to specify init_logging=False
+        rr.init("Blueprint", init_logging=False, add_to_app_default_blueprint=args.append_default, spawn=True)
+    else:
+        rr.init("Blueprint", add_to_app_default_blueprint=args.append_default, spawn=True)
```
Comment (teh-cmc) [inline examples/python/blueprint/main.py:25]: The fact that the demo itself is named `Blueprint` leads to a lot of confusion when looking at the menus:
```suggestion
    if args.blueprint_only:
        # If only using blueprint, it's important to specify init_logging=False
        rr.init("Blueprint demo", init_logging=False, add_to_app_default_blueprint=args.append_default, spawn=True)
    else:
        rr.init("Blueprint demo", add_to_app_default_blueprint=args.append_default, spawn=True)

Comment (teh-cmc) [inline examples/python/blueprint/main.py:25]: Not starting the blueprint recording at all when `--skip-blueprint` is set makes experimenting with the demo much more user-friendly:
```suggestion
        rr.init("Blueprint demo", init_blueprint=not args.skip_blueprint, add_to_app_default_blueprint=args.append_default, spawn=True)

**Comment by teh-cmc** [comment]
Watching the memory panel makes the blueprint indefinitely grow in size, which makes sense I guess since the zoom/transform state of the plot is changing but... yeah.

It is weird though that even if I stopped the auto-follow mode, it still grows, but I'm not familiar enough with how the plot handles its state.


https://github.com/rerun-io/rerun/assets/2910679/2097cf3e-87c5-49d7-ac86-1603257068d0

File: crates/re_viewer/src/app.rs [hunk 18]
```diff
@@ -413,9 +406,48 @@ impl App {
...
                     store_stats,
+                    blueprint_stats,
                 );
             });
     }
+
+    // Materialize the blueprint from the DB if the selected blueprint id isn't the default one.
+    fn load_or_create_blueprint(
+        &mut self,
+        this_frame_blueprint_id: &RecordingId,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer/src/app.rs:417]: To make it clear it's the blueprint being used for this frame (which is currently being processed), in particular, to contrast with the value of `selected_rec_id` which might be mutated during the frame.

File: crates/re_data_store/src/lib.rs [hunk 19]
```diff
@@ -10,11 +10,13 @@ pub mod entity_properties;
 pub mod entity_tree;
 mod instance_path;
 pub mod log_db;
+mod util;
 
 pub use entity_properties::*;
 pub use entity_tree::*;
 pub use instance_path::*;
 pub use log_db::LogDb;
+pub use util::*;
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_data_store/src/lib.rs:19]: For stuff like this I'm always going to default to matching the existing patterns established in the files.

File: crates/re_sdk/src/lib.rs [hunk 20]
```diff
@@ -162,6 +164,7 @@ pub fn new_recording_info(
             rustc_version: env!("RE_BUILD_RUSTC_VERSION").into(),
             llvm_version: env!("RE_BUILD_LLVM_VERSION").into(),
         },
+        recording_type: re_log_types::RecordingType::Data,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_sdk/src/lib.rs:170]: I don't want to force a rust-API change on this one until we know we have the naming right.  I'm still not really a fan of "data recording" vs "blueprint recording".  My preference on user-facing terminology is still "Recording" vs "Blueprint".

File: crates/re_viewer/src/blueprint_components/panel.rs [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+pub struct PanelState {
+    pub expanded: bool,
+}
+
+impl PanelState {
+    // TODO(jleibs): Would be nice if this could be a const EntityPath but making
+    // the hash const is a bit of a pain.
+    pub const BLUEPRINT_VIEW: &str = "blueprint_view";
+    pub const SELECTION_VIEW: &str = "selection_view";
+    pub const TIMELINE_VIEW: &str = "timeline_view";
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer/src/blueprint_components/panel.rs:18]: According to @nikolausWest and @emilk these are called "views" not "panels"

File: examples/python/blueprint/main.py (hunk 17 above)
Comment (jleibs) [reply examples/python/blueprint/main.py]: Can you elaborate?

init_blueprint=True is the default, so part of what I want to show here is the default behavior of initializing a blueprint but then not using it.

File: rerun_py/rerun_sdk/rerun/log/experimental/blueprint.py [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    Add a new space view to the blueprint.
+
+    Parameters
+    ----------
+    name : str
+        The name of the space view.
+    space_path : str
+        The anchor point of the space view.
+    entity_paths : List[str]
+        The entities to be shown in the space view.
```
Comment (emilk) [inline rerun_py/rerun_sdk/rerun/log/experimental/blueprint.py:103]: Instinctively as a user I want to just say "Create a space-view with these objects", e.g. `add_space_view(["rgb", "depth"])`. Name feels like it could be optional, and `space_path` is just under-explained. What does it mean?

Perhaps we should just make all arguments named for now to make it easier to move/remove them later?

```suggestion
def add_space_view(
    *,
    name: str,
    space_path: str,
    entity_paths: List[str],
    blueprint: Optional[RecordingStream] = None,
) -> None:
    """
    Add a new space view to the blueprint.

    Parameters
    ----------
    name : str
        The name of the space view.
    space_path : str
        The anchor point of the space view.
    entity_paths : List[str]
        The entities to be shown in the space view.
```

File: rerun_py/rerun_sdk/rerun/log/experimental/blueprint.py [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+from typing import List, Optional
+
+import rerun_bindings as bindings  # type: ignore[attr-defined]
+
+from rerun.recording_stream import RecordingStream
+
+
+def new_blueprint(
+    application_id: str,
+    blueprint_id: Optional[str] = None,
```
Comment (emilk) [inline rerun_py/rerun_sdk/rerun/log/experimental/blueprint.py:11]: with this many arguments, we should mandate that they are named!

```suggestion
def new_blueprint(
    application_id: str,
    *,
    blueprint_id: Optional[str] = None,
```

**Review by emilk — APPROVED** [review]
